File: 19_数学/采样/拒绝采样/478. 在圆内随机生成点.ts
class Solution {
  constructor(public radius: number, public x_center: number, public y_center: number) {}

  /**
   * 写一个在圆中产生均匀随机点的函数 randPoint
   * @summary
   * 考虑一个半径为1的圆，点落在半径为x的圆内概率为x^2，
   * 所以为了使点均匀分布，x=sqrt(U),这样概率就是U
   */
  randPoint(): number[] {
    const r = this.radius * Math.sqrt(Math.random())
    const degree = Math.random() * Math.PI * 2
    const x = this.x_center + r * Math.cos(degree)
    const y = this.y_center + r * Math.sin(degree)
    return [x, y]
  }
}

const s = new Solution(1, 0, 0)

export {}

File: 19_数学/采样/478. 在圆内随机生成点.py
from dataclasses import dataclass
from math import cos, pi, sin, sqrt
from random import random, uniform
from typing import List


@dataclass
class Solution:
    radius: float
    x_center: float
    y_center: float

    def randPoint(self) -> List[float]:
        # !在单位圆中随机生成一个点，它离圆心的距离小于等于 r 的概率为 F(r) = r^2
        randR = self.radius * sqrt(random())  # 注意这里是根号
        randArc = uniform(0, 2 * pi)
        x = self.x_center + randR * cos(randArc)
        y = self.y_center + randR * sin(randArc)
        return [x, y]

